core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java

```diff
@@ -337,8 +337,14 @@ private boolean doHandleTopologyUpdate(String cacheName, CacheTopology cacheTopo
                case CONFLICT_RESOLUTION:
                   // Ensure that this node utilises it's old partitions readConsistentHash during conflict resolution
                   // But has an updated write consistent hash which contains owners from the pre and post merge hashes
-                  unionCH = chf.union(existingTopology.getWriteConsistentHash(), cacheTopology.getPendingCH());
-                  currentCH = existingTopology.getCurrentCH();
+
+                  if (existingTopology != null) {
+                     ConsistentHash existingWriteHash = existingTopology.getPhase() == CacheTopology.Phase.NO_REBALANCE ? existingTopology.getCurrentCH() : existingTopology.getPendingCH();
+                     unionCH = chf.union(existingWriteHash, cacheTopology.getPendingCH());
+                     currentCH = existingTopology.getCurrentCH();
```

Comment:
It's necessary to utilise the current node's view of read owners so that the user only reads pre-conflict resolution values. Consider a cluster `{A,B,C,D}` with `ALLOW_READ_WRITES` that has a Key `k` with owners `{A,C}` and value `1`, the cluster then splits into `{A,B} {C,D}`. 

1. User executes `put(k,2)` on {C}
2. Merge starts and conflict resolution starts
3. User executes `read(k)` on {C}
4. Conflict resolution finishes and rebalance occurs

During stage 3, the user should receive `k=2` as the conflict resolution has not finished, therefore we don't know if the final "resolved" value will be 1 or 2. If we didn't use the old read owners of the current node, then it would be possible for the user to receive `k=1` during stage 3 and then receive `k=2` when executing `get(k)` after conflict resolution.

Reply:
`OperationsDuringMergeConflictTest` tests a similar scenario as described above.

Reply:
Aha, I misunderstood how `Phase.CONFLICT_RESOLUTION` works: I was expecting reads to go to all the owners and perform the conflict resolution every time. I totally understand why you wouldn't want to fetch the value from all the owners, though, it could stress the cluster quite a bit!

I also agree with the goal to prevent the clients from seeing an old value after they already saw a newer value, but I think the "why" should be written explicitly in the comment above along with the "how". And I should mention that because of staggered gets we cannot actually guarantee that goal, even in a stable cluster ;)